trin-execution/src/execution.rs

```diff
@@ -123,18 +123,40 @@ impl State {
     /// This is a lot faster then process_block() as it executes the range in memory, but we won't
     /// return the trie diff so you can use this to sync up to the block you want, then use
     /// `process_block()` to get the trie diff to gossip on the state bridge
-    pub async fn process_range_of_blocks(
-        &mut self,
-        start: u64,
-        end: u64,
-    ) -> anyhow::Result<RootWithTrieDiff> {
+    pub async fn process_range_of_blocks(&mut self, end: u64) -> anyhow::Result<RootWithTrieDiff> {
+        let start = self.execution_position.next_block_number();
+        ensure!(
+            end >= start,
+            "End block number {} is less than start block number {}",
+            end,
+            start
+        );
+
+        // If we are starting from the beginning, we need to initialize the genesis state
+        if start == 0 {
```

Comment:
I still think that you can/should put genesis logic into BlockExecutor (we can consider this initialization the execution of block zero). We would have to refactor a bit how it works, but nothing special.

In this file, you would remove this check, and handle block zero the same way any other block is handled (just call `block_executor.execute_block(&block)?`).

And in the block_executor, you would have something like this:

```rust
fn process_genesis(&mut self) -> anyhow::Result<()> {
    let genesis_file = ...;
    let genesis: GenesisConfig = ...;
    self.increment_balances(genesis.alloc.iter().map(|(address, alloc)| (address, alloc.balance)))
}

pub fn execute_block(&mut self, block: &ProcessedBlock) -> anyhow::Result<()> {
    ...
    // somewhere in this function, either at the start or next to DAO_FORK
    if block.header.number == 0 {
        self.process_genesis()?;
    }
    ...
}
```

Note: I think it does, but I'm not 100% sure that `increment_balances` would work on non-existing accounts. If it doesn't, you can use [self.evm.db_mut().insert_account](https://reth.rs/docs/reth/revm/revm/struct.State.html#method.insert_account)

Reply:
```
        for (address, alloc_balance) in genesis.alloc {
            self.evm
                .db_mut()
                .insert_account(address, AccountInfo::from_balance(alloc_balance.balance));
        }

        // self.increment_balances(
        //     genesis
        //         .alloc
        //         .into_iter()
        //         .map(|(address, alloc)| (address, alloc.balance.to::<u128>())),
        // )
```

I tried both of these and they caused the wrong state root to be generated,

```
        for (address, alloc_balance) in genesis.alloc {
            let address_hash = keccak256(address);
            let mut account = RocksAccount::default();
            account.balance += alloc_balance.balance;
            self.evm.db().database.trie.lock().insert(
                address_hash.as_ref(),
                &alloy_rlp::encode(AccountStateInfo::from(&account)),
            )?;
            self.evm
                .db()
                .database
                .db
                .put(address_hash, alloy_rlp::encode(account))?;
        }
```

So I am just going to do this ^. I think this suggestion is an unintended usecase of the EVM cache tbh, and it isn't worth debugging this